Update EIP dynamic router router (#803)
wso2/docs-mi: en/docs/learn/enterprise-integration-patterns/message-routing/dynamic-router.md

> This sample scenario demonstrates a router that takes an incoming request and decides which back-end service to transmit the message to. To make that decision, it uses a property in the message itself, very much like the Content-Based Router. However, it can also cross-check a registry entry to see if a specific endpoint accepts messages with that property. This approach allows you to reconfigure the router when registry entries change.

Reviewer: ```suggestion
This sample scenario demonstrates a router that takes an incoming request and decides which back-end service to transmit the message to. To make that decision, it uses a property in the message itself, similar to a Content-Based Router. However, it can also cross-check a registry entry to see if a specific endpoint accepts messages with that property. This approach allows you to reconfigure the router when registry entries change.
```

| Dynamic Rule Base             | Local Registry                            | 
| Dynamic Router                | Switch Mediator                           | 
| A, B, C                       | Simple Stock Quote Services               |

## Synapse configuration of the artifacts

=== "Proxy Service"
    ```xml
    <?xml version="1.0" encoding="UTF-8"?>
    <proxy name="DynamicRouterProxy" startOnLoad="true" transports="http https" xmlns="http://ws.apache.org/ns/synapse" statistics="disable" trace="disable">
        <target>
            <inSequence>
                <log category="INFO" level="full"/>
                <switch source="get-property('To')">
                    <case regex="http://localhost:9000.*">
                        <filter xmlns:m0="http://services.samples" xpath="get-property('ConfigA') = //m0:getQuote/m0:request/m0:symbol/text()">
                            <then>
                                <call>
                                    <endpoint key="StockQuoteServiceEP1"/>
                                </call>
                            </then>
                            <else>
                                <log category="INFO" level="custom">

>                                     <property name="MESSAGE" value="Registry Value Doesn't Matched"/>

Reviewer 2: ```suggestion
                                    <property name="MESSAGE" value="Registry Value Doesn't Match"/>
```

- **endpoints**: Each of the three endpoints (`StockQuoteServiceEP1`, `StockQuoteServiceEP2`, `StockQuoteServiceEP3`) corresponds to a different service URL. Depending on the routing logic, the message is sent to one of these endpoints.

- **else block**: If the message does not match the expected symbol, a log message is generated indicating that the registry value didn't match, and the message is not forwarded to the service.

## Set up the sample scenario

Follow the below instructions to simulate this sample scenario.

{!includes/eip-set-up.md!}

3. Download the [backend service](https://github.com/wso2-docs/WSO2_EI/blob/master/Back-End-Service/axis2Server.zip).

4. Extract the downloaded zip file.

5. Open a terminal, and navigate to the `axis2Server/bin/` directory inside the extracted folder.

6. Execute the following command to start the axis2server with the SimpleStockQuote backend service:

    === "On MacOS/Linux/CentOS"   
          ```bash
          sh axis2server.sh
          ```
    === "On Windows"                
          ```bash
          axis2server.bat
          ```

> 7. Navigate to the `MI_HOME/bin/` directory and start the `tcpmon` application. 

Reviewer 2: ```suggestion
7. Navigate to the `<MI_HOME>/bin/` directory and start the `tcpmon` application. 
```

> 8. In `tcpmon` application, navigate to **Admin** tab. Add listeners to ports `9001` and `9002`, for each listener set the **target hostname** to `localhost` and **target port** to `9000` in each instance.

Reviewer 2: ```suggestion
8. In the `tcpmon` application, navigate to the **Admin** tab. Add listeners to ports `9001` and `9002`, for each listener set the **target hostname** to `localhost` and **target port** to `9000` in each instance.
```

        <wsa:Action>urn:getQuote</wsa:Action>
    </soapenv:Header>
    <soapenv:Body>
        <ser:getQuote>
            <ser:request>
                <ser:symbol>bar</ser:symbol>
            </ser:request>
        </ser:getQuote>
    </soapenv:Body>
    </soapenv:Envelope>
    ```

    ```xml
    POST http://localhost:8290/services/DynamicRouterProxy

    Accept-Encoding: gzip,deflate
    Content-Type: text/xml;charset=UTF-8
    Connection: Keep-Alive
    SOAPAction: "urn:getQuote"
    To: http://localhost:9002/services/SimpleStockQuoteService

    <soapenv:Envelope xmlns:soapenv="http://schemas.xmlsoap.org/soap/envelope/" xmlns:ser="http://services.samples" xmlns:xsd="http://services.samples/xsd" xmlns:wsa="http://www.w3.org/2005/08/addressing">
    <soapenv:Header>
        <wsa:To>http://localhost:9002/services/SimpleStockQuoteService</wsa:To>
        <wsa:Action>urn:getQuote</wsa:Action>
    </soapenv:Header>
    <soapenv:Body>
        <ser:getQuote>
            <ser:request>
                <ser:symbol>WSO2</ser:symbol>
            </ser:request>
        </ser:getQuote>
    </soapenv:Body>
    </soapenv:Envelope>
    ```

> 2. Send more requests changing `foo`, `bar`, and `WSO2` to other values.

Reviewer: Better to mention the parameter name `symbol` instead of the values

> 2. Send more requests changing `foo`, `bar` and `WSO2` to other values.

Reviewer 2: ```suggestion
2. Send more requests changing `foo`, `bar`, and `WSO2` to other values.
```

## Analyze the output

> 1. When you send the requests, you can view outputs like this. Also, you can see in `tcpmon` that you received a request for the relevant port.

Reviewer: ```suggestion
1. When you send the requests with the `symbol` value set to one of `foo`, `bar`, or `WSO2`, you will receive outputs like the following. Also, you can see in `tcpmon` that you received a request for the relevant port.
```

                    <ax21:change>3.9782739749100493</ax21:change>
                    <ax21:earnings>12.142627905802117</ax21:earnings>
                    <ax21:high>-142.69159439461524</ax21:high>
                    <ax21:last>145.1810442343239</ax21:last>
                    <ax21:lastTradeTimestamp>Tue Aug 20 11:48:05 IST 2024</ax21:lastTradeTimestamp>
                    <ax21:low>152.09105015765533</ax21:low>
                    <ax21:marketCap>3.707834504517188E7</ax21:marketCap>
                    <ax21:name>foo Company</ax21:name>
                    <ax21:open>151.25901812607216</ax21:open>
                    <ax21:peRatio>-19.743676192563544</ax21:peRatio>
                    <ax21:percentageChange>2.4263061719499506</ax21:percentageChange>
                    <ax21:prevClose>163.9642193925109</ax21:prevClose>
                    <ax21:symbol>foo</ax21:symbol>
                    <ax21:volume>19405</ax21:volume>
                </ns:return>
            </ns:getQuoteResponse>
        </soapenv:Body>
    </soapenv:Envelope>
    ```

> 2. When you change the `symbol` this log will print on the WSO2 MI console.

Reviewer: ```suggestion
2. When you change the symbol to a different value, this log will be printed to the WSO2 MI console.
```
